### PR TITLE
Small logic cleanups

### PR DESCRIPTION
**Pull request by haileyok**
## Test Plan

- Test edge cases inside of the Storybook. Press `Open Tester` to see the various edge cases and watch the console logs.
- Composer
	- Type some junk
	- Attempt to close
	- On Android, attempt to use the back button now
- Hide Post
	 - On Android, attempt to use the back button now
- Report Post
    - Again, try to use the back button on Android (I usually just do a trash report to Beans Labeler for testing, lol)
- Just swipe to close a dialog. Everything should still work.
- Use the mute words dialog
    - Delete a mute word, and make sure the back button is working

### CODE REVIEW

File: src/components/Dialog/index.web.tsx [hunk 1]
```diff
@@ -113,17 +110,15 @@ export function Outer({
                   gtMobile ? a.p_lg : a.p_md,
                   {overflowY: 'auto'},
                 ]}>
-                {isVisible && (
-                  <Animated.View
```
Comment (haileyok) [inline src/components/Dialog/index.web.tsx:117]: Do we need this `isVisible`? See above

File: src/components/Dialog/index.web.tsx [hunk 2]
```diff
@@ -33,20 +33,17 @@ export function Outer({
...
     setDialogIsOpen(control.id, true)
+    setIsOpen(true)
   }, [setIsOpen, setDialogIsOpen, control.id])
 
   const close = React.useCallback<DialogControlProps['close']>(
     cb => {
       setDialogIsOpen(control.id, false)
-      setIsVisible(false)
       setIsOpen(false)
-      setIsVisible(true)
```
Comment (haileyok) [inline src/components/Dialog/index.web.tsx:49]: `isVisible` has a default value of `true` and here we never actually would have changed this to `false`.

Comment (haileyok) [reply src/components/Dialog/index.web.tsx:49]: I also can't see a visual difference after removing this in the web version, so unsure if we need it. We're conditionally rendering in the dialog once `isOpen` changes to true, so we still get the `Animated.View` `entering` animations.

File: src/components/Dialog/index.tsx [hunk 3]
```diff
@@ -105,19 +104,34 @@ export function Outer({
...
+    if (typeof cb === 'function') {
       closeCallback.current = cb
     }
-    // initiates a close animation, the actual "close" happens in `onCloseInner`
     sheet.current?.close()
   }, [])
 
+  // This is the actual thing we are doing once we "confirm" the dialog. We want the dialog's close animation to
+  // happen before we run this. It is passed to the `BottomSheet` component.
+  const onCloseAnimationComplete = React.useCallback(() => {
```
Comment (haileyok) [inline src/components/Dialog/index.tsx:117]: Calling this `onCloseAnimationComplete` I think gives us a better sense of the purpose of this function. Too many `onClose`s gets a bit confusing I think.

File: src/components/Dialog/index.tsx [hunk 4]
```diff
@@ -105,19 +104,34 @@ export function Outer({
...
-        logger.error(
-          `Dialog close was passed multiple callbacks, you shouldn't do that`,
-        )
-      }
+    if (typeof cb === 'function') {
       closeCallback.current = cb
     }
-    // initiates a close animation, the actual "close" happens in `onCloseInner`
     sheet.current?.close()
   }, [])
```
Comment (haileyok) [inline src/components/Dialog/index.tsx:113]: Do we actually care if the callback changes? Or can it? From what I can tell if we call `close` then the dialog's lifetime is going to end and there won't be another chance for this to get called. Unless we are worried about `control.close()` getting called multiple times for some reason?

File: src/components/Dialog/index.web.tsx [hunk 5]
```diff
@@ -135,7 +130,7 @@ export function Outer({
                       minHeight: web('calc(90vh - 36px)') || undefined,
                     },
                   ]}>
-                  {isVisible ? children : null}
```
Comment (haileyok) [inline src/components/Dialog/index.web.tsx:138]: Same here

File: src/components/Dialog/index.tsx (hunk 4 above)
Comment (gaearon) [reply src/components/Dialog/index.tsx:113]: In principle we should take multiple calls into account since there's no mechanism to prevent those, and if they happen in different parts of the codebase (which at first might be independent and then due to another change might start firing together), each of them may have an expectation that the callback will fire.

File: src/components/Dialog/index.web.tsx (hunk 2 above)
Comment (gaearon) [reply src/components/Dialog/index.web.tsx:49]: State calls during events should be batched so I'm not sure there'd be a difference either way.

File: src/components/Dialog/index.tsx (hunk 4 above)
Comment (haileyok) [reply src/components/Dialog/index.tsx:113]: Okay. Perhaps I am misunderstanding what is going on outside of here, but I would imagine that we wouldn't actually hit this case. Let me look some more. I'd rather that just not be a possibility at all.

File: src/components/Dialog/index.tsx [hunk 6]
```diff
@@ -105,19 +105,36 @@ export function Outer({
...
-      }
-      closeCallback.current = cb
+    if (typeof cb === 'function') {
+      closeCallbacks.current?.push(cb)
+    }
+
+    if (!isClosing.current) {
+      sheet.current?.close()
+      isClosing.current = true
     }
```
Comment (haileyok) [inline src/components/Dialog/index.tsx:117]: Queue callbacks so that we call each one, incase there are multiple.

File: src/components/Dialog/index.tsx [hunk 7]
```diff
@@ -105,19 +105,36 @@ export function Outer({
...
   const close = React.useCallback<DialogControlProps['close']>(cb => {
-    if (cb && typeof cb === 'function') {
-      if (closeCallback.current) {
-        logger.error(
-          `Dialog close was passed multiple callbacks, you shouldn't do that`,
-        )
-      }
-      closeCallback.current = cb
+    if (typeof cb === 'function') {
+      closeCallbacks.current?.push(cb)
```
Comment (gaearon) [inline src/components/Dialog/index.tsx:112]: Why is it nullable?

File: src/components/Dialog/index.tsx [hunk 8]
```diff
@@ -105,19 +105,36 @@ export function Outer({
...
+    }
+
+    if (!isClosing.current) {
+      sheet.current?.close()
+      isClosing.current = true
```
Comment (gaearon) [inline src/components/Dialog/index.tsx:117]: nit: might want to move this line up in case close() throws for whatever reason

File: src/components/Dialog/index.tsx [hunk 9]
```diff
@@ -105,19 +105,36 @@ export function Outer({
...
+    // This removes the dialog from our list of stored dialogs. Not super necessary on iOS, but on Android this
+    // tells us that we need to toggle the accessibility overlay setting
+    setDialogIsOpen(control.id, false)
+    setOpenIndex(-1)
+
+    if (closeCallbacks.current && closeCallbacks.current.length > 0) {
+      for (const cb of closeCallbacks.current) {
+        try {
+          cb()
+        } catch {}
```
Comment (gaearon) [inline src/components/Dialog/index.tsx]: should not silence them — need to store at least the first one and rethrow at the end of the function

File: src/components/Dialog/index.tsx (hunk 7 above)
Comment (haileyok) [reply src/components/Dialog/index.tsx:112]: oops

File: src/components/Dialog/index.tsx [hunk 10]
```diff
@@ -96,28 +97,55 @@ export function Outer({
...
+  // happen before we run this. It is passed to the `BottomSheet` component.
+  const onCloseAnimationComplete = React.useCallback(() => {
+    // This removes the dialog from our list of stored dialogs. Not super necessary on iOS, but on Android this
+    // tells us that we need to toggle the accessibility overlay setting
+    setDialogIsOpen(control.id, false)
+    setOpenIndex(-1)
+
+    callQueuedCallbacks()
+    onClose?.()
+    isClosing.current = false
```
Comment (gaearon) [inline src/components/Dialog/index.tsx:146]: Set important flags before calling into external code so that our internal state doesn't get corrupted by external code throwing.

File: src/components/Dialog/index.web.tsx [hunk 11]
```diff
@@ -33,24 +33,25 @@ export function Outer({
...
-      setIsVisible(true)
 
       try {
         if (cb && typeof cb === 'function') {
-          cb()
+          // This timeout ensures that the callback runs at the same time as it would on native. I.e.
+          // console.log('Step 1') -> close(() => console.log('Step 3')) -> console.log('Step 2')
+          // This should always output 'Step 1', 'Step 2', 'Step 3', but without the timeout it would output
+          // 'Step 1', 'Step 3', 'Step 2'.
+          setTimeout(cb)
```
Comment (haileyok) [inline src/components/Dialog/index.web.tsx:54]: good callout on this one, `setTimeout` gives us the expected result.

File: src/components/Dialog/index.web.tsx (hunk 2 above)
Comment (estrattonbailey) [reply src/components/Dialog/index.web.tsx:49]: Oh ya know what, this is leftovers from an exit animation we removed, def safe to remove

File: src/components/Dialog/index.tsx [hunk 12]
```diff
@@ -96,28 +96,49 @@ export function Outer({
...
+    closeCallbacks.current = []
+  }, [])
+
   const open = React.useCallback<DialogControlProps['open']>(
     ({index} = {}) => {
       setDialogIsOpen(control.id, true)
       // can be set to any index of `snapPoints`, but `0` is the first i.e. "open"
       setOpenIndex(index || 0)
+
+      callQueuedCallbacks()
```
Comment (estrattonbailey) [inline src/components/Dialog/index.tsx:117]: Do we want to call all close callbacks here, or maybe just reset the an empty array on re-open?

Comment (haileyok) [reply src/components/Dialog/index.tsx:117]: I think calling all of the old ones is the best move. I.e:

- Press save
- Close animation begins
- Somehow the dialog gets opened again before the animation ends
- ???

I think i'd expect the first thing I did to still complete personally.
